[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/badaouihakimou/machine-learning-notebooks/blob/main/02_variables_et_fonctions.ipynb)


# Variables et fonctions en Python

Les deux briques de base. Une variable garde une valeur, une fonction garde un
traitement.

Le sujet a l'air simple, et il l'est en surface. Mais trois détails expliquent
une bonne partie des bugs qu'on rencontre ensuite, et aucun ne lève d'erreur :
la différence entre `and` et `&`, le comportement des objets mutables, et les
arguments par défaut. Ils sont traités dans les sections 3, 5 et 7.

## Le plan

| Section | Le sujet |
|---|---|
| 1 | Les variables et l'affectation |
| 2 | Les types de base |
| 3 | Les opérateurs, et le piège `and` / `&` |
| 4 | Les chaînes de caractères |
| 5 | L'affectation multiple et les objets mutables |
| 6 | Les fonctions : `def` et `lambda` |
| 7 | Les arguments : nommés, par défaut, et le piège du défaut mutable |

Aucun prérequis, sinon savoir exécuter une cellule.

## 1. Les variables

Une variable associe un nom à une valeur. Le signe `=` n'est pas une égalité
mathématique mais une affectation : il range une valeur dans un nom.

In [52]:
x = 1
print(x)

x = 2  # la variable change de valeur
print(x)

1
2


La ligne suivante surprend souvent quand on vient des mathématiques.

In [53]:
x = x + 1  # se lit : prends la valeur de x, ajoute 1, range le résultat dans x
print(x)

3


En mathématiques, `x = x + 1` n'aurait aucune solution. En Python, le côté droit
est évalué d'abord, puis le résultat est rangé dans le nom de gauche.

L'ordre est toujours celui-ci : droite d'abord, gauche ensuite.

Il existe une écriture raccourcie, qui fait exactement la même chose.

In [54]:
x += 1  # équivalent à x = x + 1
print(x)

x *= 3 # x = x * 3
print(x)

x -= 2 # x = x - 2
print(x)

4
12
10


### Nommer ses variables

Les règles imposées par Python : pas d'espace, pas de chiffre en première
position, pas de tiret, et la casse compte `age` et `Age` sont deux variables
distinctes.

La convention en Python est le `snake_case` : tout en minuscules, les mots
séparés par des tirets bas.

```python
masse_totale = 80 # oui
masseTotale = 80 # convention d'autres langages
MasseTotale = 80 # réservé aux classes
```

Un conseil qui compte plus que la convention : un nom doit dire ce que la
variable contient. `x` convient pour un compteur de boucle ; pour une masse en
kilogrammes, `masse_kg` évitera une erreur d'unité dans six mois.

## 2. Les types de base

Chaque valeur a un type. `type()` le donne.

In [55]:
entier = 1
decimal = 2.5
booleen = True
texte = 'bonjour'
print(type(entier))
print(type(decimal))
print(type(decimal))
print(type(decimal))

<class 'int'>
<class 'float'>
<class 'float'>
<class 'float'>


In [56]:
for variable in [entier, decimal, booleen, texte]:
    print(f'{str(variable):<10} -> {type(variable)}')

1          -> <class 'int'>
2.5        -> <class 'float'>
True       -> <class 'bool'>
bonjour    -> <class 'str'>


Les quatre types fondamentaux :

| Type | Nom Python | Exemples |
|---|---|---|
| entier | `int` | `1`, `-42`, `1000000` |
| décimal | `float` | `2.5`, `-0.001`, `3.0` |
| booléen | `bool` | `True`, `False` |
| chaîne | `str` | `'bonjour'`, `"a"`, `''` |

Attention : `3` et `3.0` n'ont pas le même type, même s'ils valent la même chose.

### Le piège des flottants

Les nombres décimaux sont stockés en binaire, ce qui provoque de petites erreurs
d'arrondi. C'est déroutant la première fois.

In [57]:
print(0.1 + 0.2)
print(0.1 + 0.2 == 0.3)

0.30000000000000004
False


Le résultat n'est pas `0.3` mais `0.30000000000000004`, et la comparaison est
donc fausse.

Ce n'est pas un bug de Python : c'est une conséquence du codage binaire, et tous
les langages ont le même comportement. Certains nombres décimaux n'ont pas de
représentation binaire exacte, exactement comme 1/3 n'a pas d'écriture décimale
finie.

La conséquence pratique : ne jamais tester l'égalité de deux flottants. On
compare à une tolérance près.

In [59]:
print(abs((0.1 + 0.2) - 0.3) < 1e-9) # la bonne façon

True


In [60]:
import math
print(math.isclose(0.1 + 0.2, 0.3)) # ou la fonction dédiée

True


### Convertir un type

In [61]:
print(int('42') + 1) # texte -> entier
print(float('2.5') * 2) # texte -> décimal
print(str(42) + ' ans') # nombre -> texte
print(int(2.9)) # tronque, n'arrondit pas
print(round(2.9)) # arrondit

43
5.0
42 ans
2
3


`int()` tronque vers zéro, il n'arrondit pas. `int(2.9)` vaut 2, et `int(-2.9)`
vaut -2.

Et la conversion échoue si elle n'a pas de sens :

```python
int('bonjour') # ValueError
```

In [62]:
int('bonjour') # ValueError

ValueError: invalid literal for int() with base 10: 'bonjour'

## 3. Les opérateurs

### Arithmétiques

In [63]:
x = 7
y = 2

print('x + y  =', x + y)
print('x - y  =', x - y)
print('x * y  =', x * y)
print('x / y  =', x / y) # division réelle, donne toujours un float
print('x // y =', x // y) # division entière : le quotient
print('x %  y =', x % y) # modulo : le reste
print('x ** y =', x ** y) # puissance

x + y  = 9
x - y  = 5
x * y  = 14
x / y  = 3.5
x // y = 3
x %  y = 1
x ** y = 49


Les deux dernières lignes méritent l'attention.

`//` et `%` vont souvent ensemble. Le quotient et le reste permettent de placer
un élément dans une grille : c'est exactement ce qui servait à disposer les
couleurs en lignes et colonnes dans le notebook Matplotlib.

```python
ligne, colonne = i % 30, i // 30
```

Le modulo sert aussi à tester la divisibilité :

In [64]:
for n in range(1, 11):
    if n % 2 == 0:
        print(n, 'est pair')

2 est pair
4 est pair
6 est pair
8 est pair
10 est pair


### Comparaison

Ces opérateurs renvoient un booléen.

In [65]:
x = 1
y = 2.5

print('x <  y :', x < y)
print('x <= y :', x <= y)
print('x == y :', x == y) # deux signes égal : comparaison
print('x != y :', x != y) # différent de

x <  y : True
x <= y : True
x == y : False
x != y : True


Ne pas confondre `=` et `==`. Le premier affecte, le second compare. C'est une
confusion classique, mais Python lève une erreur de syntaxe assez claire quand on
se trompe.

### Logiques : le piège de la section

Python a deux jeux d'opérateurs logiques, et ils ne sont pas
interchangeables.

In [66]:
print('and / or / not  ->  opérateurs logiques')
print(False and True)
print(False or True)
print(not False)
print()
print('& / | / ^  ->  opérateurs bit à bit')
print(False & True)
print(False | True)
print(False ^ True) # ou exclusif : vrai si l'un OU l'autre, pas les deux

and / or / not  ->  opérateurs logiques
False
True
True

& / | / ^  ->  opérateurs bit à bit
False
True
True


Sur des booléens, les deux jeux donnent le même résultat. D'où la confusion :
on croit qu'ils sont équivalents.

Ils ne le sont pas. Voici deux différences qui comptent.

Différence 1 : sur des nombres, les résultats divergent.

In [67]:
print('2 and 3 =', 2 and 3) # renvoie 3
print('2  &  3 =', 2 & 3) # renvoie 2
print()
print('1 and 2 =', 1 and 2) # renvoie 2
print('1  &  2 =', 1 & 2) # renvoie 0

2 and 3 = 3
2  &  3 = 2

1 and 2 = 2
1  &  2 = 0


`and` regarde si les valeurs sont vraies ou fausses et renvoie l'une des deux
opérandes. `&` travaille sur la représentation binaire, bit par bit.

En binaire, 1 s'écrit `01` et 2 s'écrit `10`. Aucun bit n'est à 1 aux deux
positions, donc `1 & 2` vaut 0.

Différence 2 : `and` s'arrête dès qu'il peut conclure.

In [68]:
def bruyante():
    print('-> la fonction a été appelée')
    return True

print('Avec and :')
resultat = False and bruyante()

print('Avec & :')
resultat = False & bruyante()

Avec and :
Avec & :
-> la fonction a été appelée


Avec `and`, la fonction n'est jamais appelée : Python sait déjà que le résultat
sera faux, il s'arrête là. C'est ce qu'on appelle l'évaluation court-circuit.

Avec `&`, les deux côtés sont évalués systématiquement.

Cette différence est utile en pratique :

```python
if liste and liste[0] > 5:      # sûr : liste[0] n'est évalué que si liste n'est pas vide
```

Avec `&`, cette ligne planterait sur une liste vide.

### Lequel utiliser

En Python pur, toujours `and`, `or`, `not`. Ce sont les opérateurs logiques.

Avec NumPy et Pandas, toujours `&`, `|`, `~`. Là, on compare des tableaux
entiers élément par élément, et `and` ne sait pas faire il lève une erreur.

C'est exactement ce qu'on rencontre en filtrant un DataFrame :

```python
data[(data['age'] < 18) & (data['pclass'] == 3)]
```

Les parenthèses autour de chaque condition sont obligatoires, parce que `&` a une
priorité plus forte que `<` et `==`.

## 4. Les chaînes de caractères

Guillemets simples ou doubles, au choix. Les doubles sont pratiques quand le
texte contient une apostrophe.

In [69]:
prenom = 'Ammar'
nom = "Ben Ali"

print(prenom + ' ' + nom) # concaténation
print(prenom * 3) # répétition
print(len(prenom)) # longueur
print(prenom.upper())
print(prenom[0]) # premier caractère, l'indice commence à 0
print(prenom[-1]) # dernier caractère

Ammar Ben Ali
AmmarAmmarAmmar
5
AMMAR
A
r


### Les f-strings

C'est la façon moderne d'insérer des valeurs dans du texte. Le `f` devant les
guillemets active l'interprétation des accolades.

In [70]:
age = 25
taille = 1.7523

print(f'{prenom} a {age} ans')
print(f'Taille : {taille:.2f} m') # 2 décimales
print(f'Dans 10 ans : {age + 10} ans') # un calcul dans les accolades

Ammar a 25 ans
Taille : 1.75 m
Dans 10 ans : 35 ans


Le `:.2f` est un format : deux décimales, notation fixe. Quelques autres :

| Format | Effet | Exemple |
|---|---|---|
| `:.3f` | trois décimales | `3.142` |
| `:>10` | aligné à droite sur 10 caractères | `     texte` |
| `:<10` | aligné à gauche | `texte     ` |
| `:.1%` | pourcentage | `42.0%` |
| `:,` | séparateur de milliers | `1,000,000` |

In [71]:
proportion = 0.4237
print(f'{proportion:.1%}')
print(f'{1234567:,}')

42.4%
1,234,567


## 5. Affectation multiple et objets mutables

Python permet d'affecter plusieurs variables sur une ligne.

In [72]:
x, y = 1, 2.5

print(x, y)

1 2.5


C'est le même déballage que dans une boucle `for a, b in ...`. Le côté droit
forme un couple, le côté gauche le déballe.

L'usage le plus élégant : échanger deux valeurs sans variable temporaire.

In [73]:
a, b = 10, 20
print('avant :', a, b)
a, b = b, a
print('après :', a, b)

avant : 10 20
après : 20 10


Le côté droit est entièrement évalué avant toute affectation, ce qui rend
l'échange possible. Dans la plupart des autres langages, il faudrait passer par
une troisième variable.

### Le piège : mutable ou non

Voici le comportement qui déroute le plus, et il n'y a aucune erreur pour
prévenir.

In [74]:
# Avec des entiers : tout se passe comme prévu
a = 5
b = a
b += 1

print('a =', a, '   b =', b)

a = 5    b = 6


In [75]:
# Avec des listes : surprise
liste_a = [1, 2, 3]
liste_b = liste_a
liste_b.append(4)

print('liste_a =', liste_a)
print('liste_b =', liste_b)

liste_a = [1, 2, 3, 4]
liste_b = [1, 2, 3, 4]


`liste_a` a changé alors qu'on n'y a pas touché.

L'explication : `liste_b = liste_a` ne copie pas la liste. Les deux noms
désignent le même objet en mémoire. Modifier l'un revient à modifier l'autre.

`id()` donne l'adresse mémoire et permet de le vérifier :

In [76]:
liste_a = [1, 2, 3]
liste_b = liste_a
liste_c = liste_a.copy() # une vraie copie

print('a et b sont le même objet :', id(liste_a) == id(liste_b))
print('a et c sont le même objet :', id(liste_a) == id(liste_c))

liste_b.append(4)
liste_c.append(99)

print()
print('liste_a =', liste_a) # modifiée par b, pas par c
print('liste_c =', liste_c)

a et b sont le même objet : True
a et c sont le même objet : False

liste_a = [1, 2, 3, 4]
liste_c = [1, 2, 3, 99]


### Pourquoi les entiers se comportent différemment

Parce qu'ils sont immuables : on ne peut pas modifier un entier, seulement en
créer un nouveau. `b += 1` fabrique un nouvel entier et le range dans `b`, sans
toucher à `a`.

| Immuables | Mutables |
|---|---|
| `int`, `float`, `bool` | `list` |
| `str` | `dict` |
| `tuple` | `set` |
| | tableaux NumPy, DataFrames Pandas |

C'est la même distinction qui explique pourquoi il faut écrire `data =
data.dropna()` en Pandas : la méthode renvoie un nouvel objet au lieu de modifier
l'original.

Et pourquoi `.copy()` était nécessaire quand on comparait trois traitements d'un
même DataFrame sans lui, les trois auraient partagé les mêmes données.

## 6. Les fonctions

Une fonction met un traitement de côté sous un nom, pour le réutiliser.

In [77]:
def energie_potentielle(masse, hauteur, g):
    E = masse * hauteur * g
    return E

resultat = energie_potentielle(80, 5, 9.81)
print(resultat, 'Joules')

3924.0 Joules


Les éléments de la définition :

- `def` ouvre la déclaration,
- le nom de la fonction,
- les paramètres entre parenthèses,
- les deux points,
- le corps, indenté,
- `return` renvoie le résultat.

L'indentation n'est pas cosmétique en Python : c'est elle qui délimite le corps
de la fonction. Une ligne non indentée en sort.

### print ou return ?

C'est une confusion très fréquente.

In [78]:
def avec_print(masse, hauteur, g):
    print(masse * hauteur * g) # affiche, ne renvoie rien

def avec_return(masse, hauteur, g):
    return masse * hauteur * g # renvoie, n'affiche rien

a = avec_print(80, 5, 9.81)
b = avec_return(80, 5, 9.81)

print()
print('a vaut :', a) # None : la fonction n'a rien renvoyé
print('b vaut :', b)

3924.0

a vaut : None
b vaut : 3924.0


`print` affiche à l'écran. `return` renvoie une valeur réutilisable.

Une fonction sans `return` renvoie `None`, ce qui rend son résultat inutilisable :

In [79]:
try:
    print(avec_print(80, 5, 9.81) * 2)
except TypeError as e:
    print('TypeError :', e)

3924.0
TypeError : unsupported operand type(s) for *: 'NoneType' and 'int'


La règle : une fonction calcule et renvoie, elle n'affiche pas. L'affichage
se fait à l'extérieur.

```python
resultat = energie_potentielle(80, 5, 9.81)
print(f'{resultat:.1f} Joules')
```

Une fonction qui affiche n'est utilisable que dans un notebook. Une fonction qui
renvoie s'utilise partout : dans un calcul, dans une boucle, dans une autre
fonction.

### Les lambda

Une lambda est une fonction en une ligne, sans nom.

In [80]:
carre = lambda x: x**2
print(carre(3), carre(4), carre(5))

# Exactement équivalent à :
def carre_bis(x):
    return x**2

print(carre_bis(3))

9 16 25
9


La lambda a un `return` implicite : le résultat de l'expression est renvoyé
automatiquement.

Elle accepte plusieurs paramètres :

In [81]:
combinaison = lambda x, y: x**2 + y
print(combinaison(3, 4))

13


### Quand utiliser une lambda

Presque jamais pour la nommer. `carre = lambda x: x**2` est moins clair qu'un
`def`, et le guide de style officiel de Python le déconseille explicitement.

Son intérêt est d'être passée en argument à une autre fonction, sans avoir à la
déclarer :

```python
sorted(passagers, key=lambda p: p['age'])
data.groupby('titre')['age'].transform(lambda s: s.fillna(s.median()))
optimize.differential_evolution(lambda p: f(p[0]), bounds=[(-10, 10)])
```

Ce sont exactement les usages rencontrés dans les notebooks Pandas et SciPy.
Écrire un `def` pour trois caractères d'expression serait plus lourd.

### Documenter

Une chaîne placée en première ligne du corps devient la documentation de la
fonction.

In [82]:
def energie_potentielle(masse, hauteur, g=9.81):
    """Calcule l'énergie potentielle de pesanteur.
    masse : en kilogrammes
    hauteur : en mètres
    g : accélération de la pesanteur, en m/s²
    Retourne l'énergie en Joules.
    """
    return masse * hauteur * g

help(energie_potentielle)

Help on function energie_potentielle in module __main__:

energie_potentielle(masse, hauteur, g=9.81)
    Calcule l'énergie potentielle de pesanteur.
    masse : en kilogrammes
    hauteur : en mètres
    g : accélération de la pesanteur, en m/s²
    Retourne l'énergie en Joules.



Cette docstring est ce qu'affiche `help()`, et ce que montre l'éditeur quand on
survole le nom de la fonction. Sur une fonction avec des unités, comme ici, elle
évite une erreur de calcul par confusion de grammes et de kilogrammes.

## 7. Les arguments

### Positionnels ou nommés

Par défaut, les arguments sont associés dans l'ordre où ils arrivent.

In [83]:
def energie(masse, hauteur, g=9.81):
    return masse * hauteur * g

print(energie(80, 5, 9.81)) # positionnel : l'ordre compte
print(energie(masse=80, hauteur=5)) # nommé : l'ordre n'a plus d'importance
print(energie(hauteur=5, masse=80)) # même résultat

3924.0
3924.0
3924.0


Nommer les arguments rend le code lisible et supprime les erreurs d'ordre. Sur
une fonction à quatre paramètres, `energie(5, 80)` a l'air correct mais calcule
autre chose.

Une contrainte : les arguments positionnels doivent précéder les nommés.

```python
energie(80, hauteur=5)      # correct
energie(masse=80, 5)        # SyntaxError
```

### Les valeurs par défaut

Un paramètre peut avoir une valeur par défaut, utilisée si l'appelant ne fournit
rien.

In [84]:
print(energie(80, 5)) # g vaut 9.81, sur Terre
print(energie(80, 5, g=1.62)) # sur la Lune
print(energie(80, 5, g=24.79)) # sur Jupiter

3924.0
648.0
9916.0


Les paramètres avec valeur par défaut doivent venir après ceux qui n'en ont
pas. `def f(g=9.81, masse)` est une erreur de syntaxe.

### Le piège du défaut mutable

Celui-ci est spectaculaire, et il ne lève aucune erreur.

In [85]:
def ajouter(element, liste=[]): # Ne jamais faire ca
    liste.append(element)
    return liste

print(ajouter(1))
print(ajouter(2))
print(ajouter(3))

[1]
[1, 2]
[1, 2, 3]


On attendrait `[1]`, `[2]`, `[3]`. On obtient `[1]`, `[1, 2]`, `[1, 2, 3]`.

L'explication : la valeur par défaut est créée une seule fois, au moment où
Python lit la définition, pas à chaque appel. Tous les appels partagent donc la
même liste, qui se remplit au fil du temps.

C'est la mutabilité de la section 5 qui frappe à nouveau. Avec un entier ou une
chaîne, immuables, le problème n'existe pas.

La solution consiste à utiliser `None` comme défaut :

In [86]:
def ajouter(element, liste=None):
    if liste is None:
        liste = [] # une liste neuve à chaque appel
    liste.append(element)
    return liste

print(ajouter(1))
print(ajouter(2))
print(ajouter(3))

[1]
[2]
[3]


La règle à retenir : jamais de liste, de dictionnaire ou d'ensemble comme
valeur par défaut. Toujours `None`, puis créer l'objet dans le corps.

### La portée des variables

Une variable créée dans une fonction n'existe qu'à l'intérieur.

In [87]:
def ma_fonction():
    interne = 42
    return interne

print(ma_fonction())

try:
    print(interne)
except NameError as e:
    print('NameError :', e)

42
NameError : name 'interne' is not defined


À l'inverse, une fonction peut lire les variables définies à l'extérieur.

In [88]:
constante = 10

def utilise_externe(x):
    return x * constante # lit constante sans qu'elle soit un paramètre

print(utilise_externe(5))

50


Ça fonctionne, mais c'est une mauvaise habitude. La fonction dépend alors d'une
variable invisible dans sa signature : si `constante` change ailleurs, le
résultat change sans qu'on comprenne pourquoi.

Une fonction devrait dépendre uniquement de ses paramètres. C'est ce qui la rend
testable et réutilisable.

```python
def utilise_parametre(x, constante=10):
    return x * constante
```

Et modifier une variable externe demande le mot-clé `global`, qui est presque
toujours le signe d'un mauvais découpage :

In [89]:
compteur = 0

def incremente_mal():
    global compteur # à éviter
    compteur += 1

def incremente_bien(compteur):
    return compteur + 1  # préférable

incremente_mal()
print('avec global :', compteur)
print('sans global :', incremente_bien(compteur))

avec global : 1
sans global : 2


## 8. Mémo

### Les trois pièges silencieux

| Situation | Ce qui se passe | La parade |
|---|---|---|
| `0.1 + 0.2 == 0.3` | vaut `False` | `math.isclose()` |
| `b = a` sur une liste | les deux noms pointent le même objet | `a.copy()` |
| `def f(x, liste=[])` | la liste est partagée entre les appels | `liste=None` |

### and et &

| | Python pur | NumPy / Pandas |
|---|---|---|
| et | `and` | `&` |
| ou | `or` | `\|` |
| non | `not` | `~` |

`and` s'arrête dès qu'il peut conclure, `&` évalue toujours les deux côtés.
Sur des nombres, ils donnent des résultats différents.

### print ou return

| | Effet | Réutilisable |
|---|---|---|
| `print(x)` | affiche | non, renvoie `None` |
| `return x` | renvoie | oui |

Une fonction calcule et renvoie. L'affichage se fait à l'extérieur.

### Ordre des paramètres

```python
def f(obligatoire, avec_defaut=valeur):
```

Les paramètres sans défaut d'abord. À l'appel, les positionnels avant les
nommés.

## 9. Exercices

**Exercice 1**

Écris une fonction `convertir_temperature(valeur, depuis='C', vers='F')` qui
convertit entre Celsius, Fahrenheit et Kelvin. Elle doit avoir une docstring et
lever une erreur claire si l'unité demandée n'existe pas.

**Exercice 2**

Écris une fonction `statistiques(nombres)` qui renvoie la moyenne, le minimum et
le maximum d'une liste, sans utiliser `numpy` ni les fonctions `min` et `max` de
Python. Renvoie les trois valeurs d'un coup et récupère-les par déballage.

**Exercice 3**

Voici une fonction qui contient le piège du défaut mutable :

```python
def enregistrer(nom, carnet={}):
    carnet[nom] = len(nom)
    return carnet
```

Montre le problème en l'appelant trois fois, explique ce qui se passe, puis
corrige-la. Vérifie ta correction avec `id()`.

## Pour continuer

La suite naturelle est les structures de données : listes, tuples, dictionnaires
et ensembles. Puis les boucles et les conditions, qui permettent d'en faire
quelque chose.

Un sujet à garder pour plus tard, mais qui vaut le détour : `*args` et
`**kwargs`, qui permettent d'écrire des fonctions acceptant un nombre variable
d'arguments. C'est ce qui explique la souplesse d'appel de fonctions comme
`plt.plot`.

In [90]:
# Exerice 1
def convertir_temperature(valeur, depuis='C', vers='F'):
    """Convertit une température entre Celsius, Fahrenheit et Kelvin.
    valeur : la température à convertir
    depuis : unité de départ, 'C', 'F' ou 'K' (défaut 'C')
    vers : unité d'arrivée, 'C', 'F' ou 'K' (défaut 'F')

    Retourne la température convertie, en float.
    Lève ValueError si une unité est inconnue ou si la température
    est en dessous du zéro absolu.
    """
    depuis = depuis.upper() # tolère 'c' comme 'C'
    vers = vers.upper()

    unites = ('C', 'F', 'K')
    if depuis not in unites:
        raise ValueError(f"Unité de départ inconnue : {depuis!r}. Attendu : {unites}")
    if vers not in unites:
        raise ValueError(f"Unité d'arrivée inconnue : {vers!r}. Attendu : {unites}")

    # Étape 1 : tout ramener en Celsius
    if depuis == 'C':
        celsius = valeur
    elif depuis == 'F':
        celsius = (valeur - 32) * 5/9
    else:
        celsius = valeur - 273.15

    if celsius < -273.15:
        raise ValueError(f"Température impossible : en dessous du zéro absolu")

    # Étape 2 : partir du Celsius vers l'unité demandée
    if vers == 'C':
        return celsius
    elif vers == 'F':
        return celsius * 9/5 + 32
    else:
        return celsius + 273.15

In [91]:
convertir_temperature(100,'C','F')

212.0

In [92]:
convertir_temperature(100,'f','c')

37.77777777777778

In [93]:
convertir_temperature(100,'F','C')

37.77777777777778

In [94]:
convertir_temperature(100,'F','c')

37.77777777777778

In [95]:
convertir_temperature(100,'f','C')

37.77777777777778

In [96]:
print(convertir_temperature(100)) # 212.0
print(convertir_temperature(0)) # 32.0
print(convertir_temperature(212, 'F', 'C')) # 100.0
print(convertir_temperature(0, 'C', 'K')) # 273.15
print(convertir_temperature(-40, 'C', 'F')) # -40.0

try:
    convertir_temperature(20, depuis='X')
except ValueError as e:
    print('Erreur :', e)

212.0
32.0
100.0
273.15
-40.0
Erreur : Unité de départ inconnue : 'X'. Attendu : ('C', 'F', 'K')


In [97]:
# Exercice 2

def statistiques(nombres):
    """Calcule moyenne, minimum et maximum d'une liste de nombres.
    Retourne un tuple (moyenne, minimum, maximum).
    Lève ValueError si la liste est vide.
    """
    if len(nombres) == 0:
        raise ValueError("La liste est vide")

    total = 0
    minimum = nombres[0] # on initialise avec une vraie valeur
    maximum = nombres[0]

    for n in nombres:
        total += n
        if n < minimum:
            minimum = n
        if n > maximum:
            maximum = n

    return total / len(nombres), minimum, maximum

In [98]:
moyenne, mini, maxi = statistiques([4, 8, 15, 16, 23, 42])
print(f'moyenne = {moyenne}, min = {mini}, max = {maxi}')

moyenne = 18.0, min = 4, max = 42


In [99]:
resultat = statistiques([4, 8, 15])
print(type(resultat)) # <class 'tuple'>
print(resultat) # (9.0, 4, 15)

<class 'tuple'>
(9.0, 4, 15)


In [100]:
print(statistiques([-10, -5, -1])) # (-5.33, -10, -1)
print(statistiques([5])) # (5.0, 5, 5)

(-5.333333333333333, -10, -1)
(5.0, 5, 5)


In [101]:
# Exercice 3
def enregistrer(nom, carnet={}):
    carnet[nom] = len(nom)
    return carnet

print(enregistrer('Ada'))
print(enregistrer('Bob'))
print(enregistrer('Zoe'))

{'Ada': 3}
{'Ada': 3, 'Bob': 3}
{'Ada': 3, 'Bob': 3, 'Zoe': 3}


In [102]:
print(enregistrer.__defaults__)
print('Adresse mémoire :', id(enregistrer.__defaults__[0]))

({'Ada': 3, 'Bob': 3, 'Zoe': 3},)
Adresse mémoire : 134423721050496


In [103]:
a = enregistrer('Tom')
print(id(a) == id(enregistrer.__defaults__[0])) # True

True


In [104]:
def enregistrer(nom, carnet=None):
    """Ajoute un nom au carnet, avec la longueur du nom comme valeur.
    Si aucun carnet n'est fourni, un nouveau dictionnaire est créé.
    """
    if carnet is None:
        carnet = {}  # créé à CHAQUE appel, dans le corps
    carnet[nom] = len(nom)
    return carnet

In [105]:
a = enregistrer('Ada')
b = enregistrer('Bob')

print(a) # {'Ada': 3}
print(b) # {'Bob': 3}
print('Même objet ?', id(a) == id(b)) # False

{'Ada': 3}
{'Bob': 3}
Même objet ? False


In [106]:
carnet = enregistrer('Ada')
carnet = enregistrer('Bob', carnet)
carnet = enregistrer('Zoe', carnet)
print(carnet) # {'Ada': 3, 'Bob': 3, 'Zoe': 3}

{'Ada': 3, 'Bob': 3, 'Zoe': 3}
